---
title: "Intro-to-vizRNASeq"
output: rmarkdown::html_vignette
author: "Bianca Pokhrel"
vignette: >
  %\VignetteIndexEntry{Intro-to-vizRNASeq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE, warning=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup}
library(vizRNASeq)
```

## Introduction

*vizRNASeq* is an R package that vizRNASeq is an R package that allows users to perform an exploratory analysis of RNASeq data and visualize their data through cluster analysis and heatmaps.

Use `help(package = "vizRNASeq")` for further details and references provided. To download *CellTyPETool*, use the following commands:

``` r
require("devtools")
devtools::install_github("bianca-pokhrel/vizRNASeq", build_vignettes = TRUE)
library("vizRNASeq")
```
To list all functions available in the package including functions from dependencies:

``` r
lsf.str("package:vizRNASeq")
```

## Instructions for Use

To run the mat-conversion function, simply have a count matrix and sample info dataframe to generate a DESeqDataSet object.

Example (using Pasilla dataset given as an example dataset in the package):

``` r
mat_conversion(cts_matrix, sample_info)
```

To run the reg-norm-res function, simply have a DESeqDataSet object as input to generate a variance stabilized dataset in the form of a DESeqDataSet object

``` r
reg_norm_res(DE_convert)
```

To run the cluster-map function, simply have a DESeqDataSet object as input to generate either a heatmap for hierarchical clustering or a visualization for k-means clustering where typeCluster can either be "hier" or "km" for hierarchical or k-means respectively and numClust is the number of clusters to generate which can range from 1-10. 

``` r
cluster_map(typeCluster ="km",DESeqObj = DE_convert, numClust = 4)
```

## References

H. Wickham. Reshaping data with the reshape package. Journal of Statistical Software, 21(12), 2007.

R Core Team (2020). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL https://www.R-project.org/.

Wickham et al., (2019). Welcome to the tidyverse. Journal of Open Source Software, 4(43), 1686, https://doi.org/10.21105/joss.01686

Wickham H (2016). ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York. ISBN 978-3-319-24277-4, https://ggplot2.tidyverse.org.

Wolfgang Huber and Alejandro Reyes (2020). pasilla: Data
package with per-exon and per-gene read counts of RNA-seq
samples of Pasilla knock-down by Brooks et al., Genome
Research 2011.. R package version 1.18.0.

Michael I. Love, S. (2020, October 27). Analyzing RNA-seq data with DESeq2.
Retrieved November 18, 2020,
from http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

Love, M.I., Huber, W., Anders, S. Moderated estimation of
fold change and dispersion for RNA-seq data with DESeq2
Genome Biology 15(12):550 (2014)

Warner, J. (n.d.). BitBio. Retrieved December 10, 2020,
from https://2-bitbio.com/2017/10/clustering-rnaseq-data-using-k-means.html

Andrzej Ole≈õ (2020). DEFormats: Differential gene expression data formats converter. R package version
1.18.0. https://github.com/aoles/DEFormats
